### PR TITLE
Generalized tensor list view

### PR DESCRIPTION
**Pull request by mzient**
* Make TensotListView composed of individual, independent pointers.
* Modify Resampling to use new approach.
* Move sample_range utility to tensor_view.h and add test.
* Modify PoC and test utilities.
* Add tests.

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: dali/kernels/tensor_view.h [hunk 1]
```diff
@@ -315,63 +410,102 @@ struct TensorListView<Backend, DataType, DynamicDimensions>
...
   TensorListView() = default;
   TensorListView(const TensorListView &) = default;
   TensorListView(TensorListView &&) = default;
   TensorListView &operator=(const TensorListView &) = default;
   TensorListView &operator=(TensorListView &&) = default;
+
+  ////////////////////////////////////////////////////////////////////////////
+  // Construction from contiguous memory
+
+  TensorListView(nullptr_t, const std::vector<TensorShape<sample_ndim>> &shapes)
```
Comment (Kh4L) [inline dali/kernels/tensor_view.h]: clang doesnt like it... maybe `std::nullptr_t` ?

File: dali/kernels/tensor_shape.h [hunk 2]
```diff
@@ -30,6 +30,21 @@ namespace kernels {
...
 
+namespace detail {
+
+template <int ndim1, int ndim2>
+struct is_compatible_ndim : std::integral_constant<bool,
+  ndim1 == ndim2 || ndim1 == DynamicDimensions || ndim2 == DynamicDimensions> {};
+
+template <int ndim1, int ndim2>
+struct check_compatible_ndim {
+  static_assert(is_compatible_ndim<ndim1, ndim2>::value, "Incompatible number of dimensions\n."
```
Comment (Kh4L) [inline dali/kernels/tensor_shape.h]: Why bothering with a type trait when you can write the condition directly here?

Comment (mzient) [reply dali/kernels/tensor_shape.h]: It guarantees consistency between `is_...` and `check_...`.

**Comment by mzient** [comment]
!build

File: dali/kernels/tensor_view.h [hunk 3]
```diff
@@ -69,9 +69,9 @@ bool ContainsCoords(const Shape &shape, const Position &pos) {
 /// @brief Calculates flat index of a given element in the tensor
 /// @remarks If pos has fewer dimensions than shape, the remaining offsets are assumed to be 0
 template <typename Shape, typename Position>
-ptrdiff_t CalcOffset(const Shape &shape, const Position &pos) {
+if_array_like<Position, ptrdiff_t> CalcOffset(const Shape &shape, const Position &pos) {
   ptrdiff_t ofs = pos[0];
-  const int pos_dim = ShapeDim(pos);
+  const int pos_dim = size(pos);
   const int shape_dim = ShapeDim(shape);
```
Comment (klecki) [inline dali/kernels/tensor_view.h]: If we already introduced `size`, you could remove `ShapeDim` and swap it for `size`.

File: dali/kernels/tensor_view.h [hunk 4]
```diff
@@ -222,77 +233,168 @@ TensorView<Backend, DataType, other_ndim> TensorViewBase<Backend, DataType, ndim
...
+  /// @brief Constructs a tensor list without specific memory
+  ///
+  /// The shape is taken from `shape` parameter and the `data`
+  /// vector is resized to num_samples and filled with null pointers.
+  TensorListViewBase(TensorListShape<sample_ndim> &&shape)  // NOLINT
+      : shape(std::move(shape))
+      , data(this->num_samples(), nullptr) {}
+
+  /// @brief Constructs a tensor list from non-contiguous memory
+  TensorListViewBase(data_pointers_t data, const TensorListShape<sample_ndim> &shape)
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/tensor_view.h]: I think there _was_ a reason, but maybe I solved in a different way.

File: dali/kernels/tensor_view.h [hunk 5]
```diff
@@ -222,77 +233,168 @@ TensorView<Backend, DataType, other_ndim> TensorViewBase<Backend, DataType, ndim
...
+  TensorListViewBase(data_pointers_t &&data, TensorListShape<sample_ndim> &&shape)
+      : shape(std::move(shape))
+      , data(std::move(data)) {}
+
+
+  /// @brief Constructs a tensor list from non-contiguous memory
+  template <typename U>
+  TensorListViewBase(const std::vector<U*> &data, const TensorListShape<sample_ndim> &shape)
+      : shape(shape)
+      , data(reinterpret_cast<const data_pointers_t&>(data)) {}
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/tensor_view.h]: I can statically validate the conversion here, although this constructor is protected and the one that invokes it does the check.

File: dali/kernels/tensor_shape.h [hunk 6]
```diff
@@ -442,6 +461,14 @@ struct TensorListShapeBase {
   constexpr bool empty() const { return size() == 0; }
   int num_samples() const { return size(); }
 
+  ptrdiff_t num_elements() const {
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/tensor_shape.h]: That's what `volume(int)` produces. We could use size_t, but for some reason, most of our values are signed anyway.

File: dali/kernels/tensor_view.h (hunk 3 above)
Comment (mzient) [reply dali/kernels/tensor_view.h]: `size` is undefined for scalar values, `ShapeDim` gives 1.

**Comment by mzient** [comment]
!build

File: dali/kernels/tensor_view.h [hunk 7]
```diff
@@ -355,7 +340,7 @@ struct TensorListViewBase {
       , data(this->num_samples(), nullptr) {}
 
   /// @brief Constructs a tensor list from non-contiguous memory
-  TensorListViewBase(data_pointers_t data, const TensorListShape<sample_ndim> &shape)
+  TensorListViewBase(const data_pointers_t &data, const TensorListShape<sample_ndim> &shape)
       : shape(shape)
       , data(std::move(data)) {}
```
Comment (klecki) [inline dali/kernels/tensor_view.h]: Remove the move here.

**Comment by mzient** [comment]
!build

**Pull request merged**